rainier-core/src/main/scala/com/stripe/rainier/ir/DataFunction.scala

```diff
@@ -0,0 +1,156 @@
+package com.stripe.rainier.ir
+
+import scala.annotation.tailrec
+
+/*
+Input layout:
+- numParamInputs param inputs
+- for 0 <= i < data.size:
+   - data[i].size first-run inputs
+   - for 0 <= j <= batchBits:
+     - for 0 <= k < data[i].size:
+        - 2^j batch inputs
+
+Output layout:
+- numOutputs data-less outputs
+- for 0 <= i < data.size:
+   - numOutputs first-run outputs
+   - for 0 <= j <= batchBits:
+      - numOutputs batch outputs
+ */
+case class DataFunction(cf: CompiledFunction,
+                        batchBits: Int,
+                        numParamInputs: Int,
+                        numOutputs: Int,
+                        data: Array[Array[Array[Double]]]) {
+
+  val numInputs = cf.numInputs
+  val numGlobals = cf.numGlobals
+
+  private val inputMultiplier = (1 << (batchBits + 1))
+  private val inputStartIndices =
+    data.map(_.size).scanLeft(numParamInputs) {
+      case (x, sz) => x + (sz * inputMultiplier)
+    }
+  require(inputStartIndices(data.size) == cf.numInputs)
+
+  private val outputMultiplier = batchBits + 2
+  private val outputStartIndices =
+    data.scanLeft(numOutputs) {
+      case (x, _) => x + (numOutputs * outputMultiplier)
+    }
+  require(outputStartIndices(data.size) == cf.numOutputs)
+
+  def apply(inputs: Array[Double],
+            globals: Array[Double],
+            outputs: Array[Double]): Unit = {
+    computeWithoutData(inputs, globals, outputs)
+    var i = 0
+    while (i < data.size) {
+      computeFirstWithData(inputs, globals, outputs, i)
+      computeRestWithData(inputs, globals, outputs, i)
+      i += 1
+    }
```

Comment:
do I see it right that `computeWithoutData` is the `prior` we discussed in the previous PR?
and:
 `computeFirstWithData` - `likelihoodA`
 `computeRestWithData` - likelihoodB`
?

Reply:
Correct.